openscenario/map_fragment/include/map_fragment/geometry.hpp

```diff
@@ -0,0 +1,306 @@
+  {
+    return {t * length, 0};
+  }
+  
+  auto getTangentVector_(double) -> Vector2d override
+  {
+    return {1, 0};
+  }
+};  // class Straight
+
+/**
+ * Arc of given radius and angle (positive to the left)
+ */
+class Arc : public ParametricCurve
+{
+  Point2d center_;
+
+public:
+  const double radius;
+  const double angle;
+
+  explicit Arc(double radius, double angle)
+    : radius(radius), angle(angle)
+  {
+    if (radius <= 0.0)
+    {
+      throw std::invalid_argument(
+        "Expected radius to be positive. Actual value:"
+        + std::to_string(radius)
+      );
+    }
+
+    if (!(0 < std::abs(angle) && std::abs(angle) < 2 * M_PI))
+    {
+      throw std::invalid_argument(
+        "Expected angle to be in range (0, 2 * PI) or (-2 * PI, 0). Actual value: "
+        + std::to_string(angle)
+      );
+    }
+
+    center_.y = angle > 0 ? radius : -radius;
+  }
+  
+private:
+  auto getPosition_(double t) -> Point2d override
+  {
+    auto theta = t * angle;
+    Point2d origin = {0, 0};
+    return center_ + rotate(origin - center_, theta);
+  }
+  
+  auto getTangentVector_(double t) -> Vector2d override
+  {
+    auto theta = t * angle;
+    Vector2d v = {1, 0};
+    return rotate(v, theta);
+  }
+};  // class Arc
+
+/**
+ * Combination of two or more component curves, chained after one another
+ */
+class CombinedCurve : public ParametricCurve
+{
+  std::vector<Transformation2d> transformations_;
+
+public:
+  const std::vector<ParametricCurve::Pointer> curves;
+
+  explicit CombinedCurve(std::vector<ParametricCurve::Pointer> const& curves)
+    : curves(curves)
+  {
+    if (curves.size() < 2)
+    {
+      throw std::invalid_argument(
+        "Number of curves to be combined must be two or more. Actual number: "
+        + std::to_string(curves.size())
+      );
+    }
+
+    Transformation2d current_transformation;
+
+    for (const auto& curve : curves)
+    {
+      transformations_.push_back(current_transformation);
+      
+      Point2d origin = {0, 0};
+      Transformation2d local_transformation = {
+        curve->getPosition(1.0) - origin,
+        calculateAngle(curve->getTangentVector(1.0))
+      };
+
+      current_transformation = chainTransformations(current_transformation,
+                                                    local_transformation);
+    }
+  }
+
+private:
+  auto getPosition_(double t) -> Point2d override
+  {
+    auto [curve_id, curve_t] = getCurveIdAndParameter(t);
+    auto curve = curves[curve_id];
+    auto transformation = transformations_[curve_id];
+
+    return applyTransformation(curve->getPosition(curve_t), transformation);
+  }
+
+  auto getTangentVector_(double t) -> Vector2d override
+  {
+    auto [curve_id, curve_t] = getCurveIdAndParameter(t);
+    auto curve = curves[curve_id];
+    auto transformation = transformations_[curve_id];
+
+    return applyTransformation(curve->getTangentVector(curve_t), transformation);
+  }
```

Comment:
It's a different story if it's a mutator, but is there any reason why the accessor has to be private? Otherwise, expose the member function unless exposing this member function would put the class at risk. Also, make immutable member functions such as accessors [const-qualified member functions](https://en.cppreference.com/w/cpp/language/member_functions#Member_functions_with_cv-qualifiers).

Reply:
It's because of how the base class (`ParametricCurve`) was designed. I wanted methods exposed as part of public interface (`getPosition`, `getTangentVector`) to always validate the argument (it is defined to always be between 0 and 1). However, it seemed redundant to implement that validation in each implementing class separately.

This can be seen in [this part](https://github.com/tier4/scenario_simulator_v2/pull/1108/files#diff-4724e2fd23efc14d8f10fc404fb96fbeced54e0bc8e19f55be270d46d82acf4dR105-R123) of the source code. For example:

```C++
  /**
   * Calculate curve point for given parameter t ∈ [0, 1]
   */
  auto getPosition(double t) const -> Point2d
  {
    validateCurveParameterOrThrow(t);
    return getPosition_(t);
  }
```

By making the virtual method `getPosition_` private, on one hand implementing classes do not need to worry about validation and can already assume the argument is within allowed range (this removes repeatability in source code). On the other hand, validation is still enforced in the public interface.

Reply:
Also, I think we differ in our philosophy regarding information hiding. You seem to prefer making members public unless there is a strong reason not to. And I always approached it from the other side: if something doesn't need to be public, do not expose it. Define a clear interface with known intended usage and hide everything else (having a limited interface limits impact on the users when some internal stuff changes).

So if this is your preference, I can make these methods public as well. I think I would then also need to rename them, for example to something like `getPositionWithArgumentValidated`.

Reply:
Got it.
As for the class in this pull request, you can design it based on your philosophy.

However, please rename `getPosition_` to `getPositionWithArgumentValidated`. In other words, give names that are easy for readers to read, even for private member functions, even if they are public (for all member functions). I'm often tempted to create auxiliary private members with an underscore or "aux'' at the end, but there's always a more descriptive and accurate name.
Descriptive names are not only for other contributors who will touch the code we write, but also for ourselves in the months or years to come.

Reply:
[Done](https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/beddce6643947ffee31f75592fa30db0a6704039), although I settled for `getPositionWithParameterValidated`. I think it better reflects the nature of the variable being validated than my original suggestion (it is the parameter of the parametric curve). Hope that's OK.